### PR TITLE
[Online Server] Chat Api for streaming and not streaming response

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

This PR adds chat api for online server

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

image

### CODE REVIEW

File: colossalai/inference/server/chat_service.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+        if stream == "true":
+            return self.chat_completion_stream_generator(request, request_dict, result_generator, request_id)
+        else:
+            return await self.chat_completion_full_generator(request, request_dict, result_generator, request_id)
+
+    async def chat_completion_stream_generator(self, request, request_dict, result_generator, request_id: int):
+        # Send first response for each request.n (index) with the role
+        role = self.get_chat_request_role(request, request_dict)
+        n = request_dict.get("n", 1)
```
Comment (yuehuayingxueluo) [inline colossalai/inference/server/chat_service.py:55]: What does 'n' mean here?

File: tests/test_infer/test_server.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            "--model",
+            "/home/chenjianghai/data/llama-7b-hf",
+        ]
+    )
+    ray.get(server_runner.ready.remote())
+    yield server_runner
+    ray.shutdown()
+
+
+async def test_completion(server):
```
Comment (yuehuayingxueluo) [inline tests/test_infer/test_server.py:62]: Attach the testing screenshot of this PR.

File: colossalai/inference/server/chat_service.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            if last_msg_content:
+                for i in range(n):
+                    choice_data = ChatCompletionResponseStreamChoice(
+                        index=i, message=DeltaMessage(content=last_msg_content)
+                    )
+                    data = choice_data.model_dump_json(exclude_unset=True)
+                    yield f"data: {data}\n\n"
+
+        # Send response for each token for each request.n (index)
+        [""] * n
```
[the remark this answers is not in the document]
Comment (CjhHa1) [reply colossalai/inference/server/chat_service.py:81]: `n` is the number of response the user want to receive, defaults to 1

File: tests/test_infer/test_server.py (hunk 2 above)
Comment (CjhHa1) [reply tests/test_infer/test_server.py:62]: done

File: colossalai/inference/server/chat_service.py (hunk 1 above)
Comment (CjhHa1) [reply colossalai/inference/server/chat_service.py:55]: `n` is the number of response the user want to receive, defaults to 1

File: colossalai/inference/server/chat_service.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (CjhHa1) [reply colossalai/inference/server/chat_service.py]: done

**Review by yuehuayingxueluo — APPROVED** [review]

**Pull request merged**